functions/src/api/lib/validation.ts

```diff
@@ -0,0 +1,134 @@
+  constructor(message: string, validationErrors: yup.ValidationError) {
+    super(message);
+    this.validationError = validationErrors;
+  }
+
+  validationError: yup.ValidationError;
+}
+
+export class ResponseValidationError extends Error {
+  constructor(message: string, validationError: yup.ValidationError) {
+    super(message);
+    this.validationError = validationError;
+  }
+
+  validationError: yup.ValidationError;
+}
+
+type DefaultSchema = yup.AnySchema<unknown, unknown, unknown, ''>;
+interface ValidatedState<
+  TBody extends yup.Schema = DefaultSchema,
+  TQuery extends yup.Schema = DefaultSchema,
+  TResponse extends yup.Schema = DefaultSchema,
+> extends Koa.DefaultState {
+  body: yup.Asserts<TBody>;
+  query: yup.Asserts<TQuery>;
+  response: yup.Asserts<TResponse>;
+}
+
+type Validator<
+  TBody extends yup.Schema,
+  TQuery extends yup.Schema,
+  TResponse extends yup.Schema,
+> = {
+  body: TBody;
+  query: TQuery;
+  response: TResponse;
+};
+
+type ValidatorOptions<
+  TBody extends yup.Schema = DefaultSchema,
+  TQuery extends yup.Schema = DefaultSchema,
+  TResponse extends yup.Schema = DefaultSchema,
+> = {
+  [k in keyof Partial<
+    Validator<TBody, TQuery, TResponse>
+  >]: yup.ValidateOptions;
+};
+
+const defaultOptions: ValidatorOptions = {
+  body: {
+    stripUnknown: true,
+  },
+  query: {
+    stripUnknown: true,
+  },
+  response: {
+    stripUnknown: true,
+  },
+};
+
+const validation =
+  <
+    TBody extends yup.Schema,
+    TQuery extends yup.Schema,
+    TResponse extends yup.Schema,
+  >(
+    validator: Partial<Validator<TBody, TQuery, TResponse>>,
+    options: ValidatorOptions = defaultOptions,
+  ): Koa.Middleware<ValidatedState<TBody, TQuery, TResponse>> =>
+  async (ctx: Koa.Context, next: Koa.Next) => {
+    try {
+      if (validator.body) {
+        ctx.state.body = await validator.body.validate(
+          ctx.request.body,
+          options.body,
+        );
+      }
+
+      if (validator.query) {
+        ctx.state.query = await validator.query.validate(
+          ctx.request.query,
+          options.query,
+        );
+      }
+    } catch (error) {
+      if (error instanceof yup.ValidationError) {
+        throw new RequestValidationError('Request validation failed', error);
+      } else {
+        throw error;
+      }
+    }
+
+    await next();
+
+    try {
+      if (ctx.status === 200 && !isEmpty(ctx.body) && validator.response) {
+        ctx.state.response = await validator.response.validate(
+          ctx.body,
+          options.response,
+        );
+        ctx.body = ctx.state.response;
+      }
+    } catch (error) {
+      if (error instanceof yup.ValidationError) {
+        throw new ResponseValidationError('Response validation failed', error);
+      } else {
+        throw error;
+      }
+    }
+  };
+
+export interface ResponseContext extends Koa.Context {
+  state: ValidatedState;
+}
+
+export const assertValidatedResponse =
```

Comment:
Shouldn't this assert request validation too? Even if the request might be empty

Reply:
I had that as a first thought but couldn't get it to run in the correct stage. Then I thought that request validation is a bit more explicit since we want to get the types in the handlers so I removed the assertion

Reply:
Got it. Then I think we should either manipulate the incoming data (and not `ctx.state`) or assert request schemas. Otherwise I see us create endpoints without request validation and/or not picking data from `ctx.state`.